### PR TITLE
border mode 'same'

### PR DESCRIPTION
**Pull request by isaak-l**
Added border mode 'same' and padding support for 2D and 3D spatial convolutions without CUDNN.

### CODE REVIEW

**Comment by isaak-l** [comment]
Note that this works for strides = 1
I'm still unclear on how and if we want to support 'same' for strides != 1.

File: DeepFried2/layers/SpatialConvolution.py [hunk 1]
```diff
@@ -23,13 +23,38 @@ def __init__(self, n_input_plane, n_output_plane, k_w, k_h, d_w=1, d_h=1, with_b
...
+            mode = 'valid'
+            pad_w, pad_h = self.border_mode
+            padded_input = df.T.zeros((symb_input.shape[0],
+                                       symb_input.shape[1],
+                                       symb_input.shape[2]+2*pad_w,
+                                       symb_input.shape[3]+2*pad_h))
+            symb_input = df.T.set_subtensor(padded_input[:,
+                                                         :,
+                                                         pad_h:symb_input.shape[2]+pad_w,
+                                                         pad_w:symb_input.shape[3]+pad_h], symb_input)
```
Comment (lucasb-eyer) [inline DeepFried2/layers/SpatialConvolution.py]: I believe this is a `h` <-> `w` mixup here.

Edit: In addition, `h` comes before `w`.

File: DeepFried2/layers/SpatialConvolution.py [hunk 2]
```diff
@@ -23,13 +23,38 @@ def __init__(self, n_input_plane, n_output_plane, k_w, k_h, d_w=1, d_h=1, with_b
...
+                                            image_shape=(None, self.n_input_plane) + (self.imshape or (None, None)),
+                                            filter_shape=self.w_shape,
+                                            border_mode=mode,
+                                            subsample=(self.d_h, self.d_w)
         )
 
+        if self.border_mode == 'same':
+            pad_w = (self.k_w - 1) // 2
+            pad_h = (self.k_h - 1) // 2
+            conv_output = conv_output[:,:,pad_w:symb_input.shape[2]+pad_w,pad_h:symb_input.shape[3]+pad_h]
```
Comment (lucasb-eyer) [inline DeepFried2/layers/SpatialConvolution.py]: Same here, I believe `h` should come before `w`.

**Comment by lucasb-eyer** [comment]
See inline comments, mainly mixup of h/w.

It looks to me like implementing a `pad` function (maybe in utils) would make both codes a lot simpler. Do you have time for this?

**Comment by isaak-l** [comment]
I added a pad function.
Hopefully this makes it clearer.

**Comment by lucasb-eyer** [comment]
Yeah, much clearer, thanks!

**Pull request merged**